kubeadm/scripts/PrepareNode.ps1

```diff
@@ -59,81 +59,97 @@ $global:PowershellArgs = "-ExecutionPolicy Bypass -NoProfile"
+
+    $StartKubeletFileContent = '$FileContent = Get-Content -Path "/var/lib/kubelet/kubeadm-flags.env"
+    $global:KubeletArgs = $FileContent.TrimStart(''KUBELET_KUBEADM_ARGS='').Trim(''"'')
+
+    $global:containerRuntime = {{CONTAINER_RUNTIME}}
+
+    if ($global:containerRuntime -eq "Docker") {
+        # Create host network to allow kubelet to schedule hostNetwork pods
+        if (-not (docker network ls -f name=host --format "{{ .ID }}")) {
+            docker network create -d nat host
+        }
+    } elseif ($global:containerRuntime -eq "containerD") {
+        # NOTE: For containerd the 0-containerd-nat.json network config template added by
+        # Install-containerd.ps1 joins pods to the host network.
+        Import-Module "c:\k\hns.psm1"
+        if (-not (Get-HnsNetwork | ? Name -eq "nat")) {
+            New-HnsNetwork -Type NAT -Name nat
+        }
+    }
 
-$env:path += ";$global:NssmInstallDirectory"
-$newPath = "$global:NssmInstallDirectory;" +
-[Environment]::GetEnvironmentVariable("PATH", [EnvironmentVariableTarget]::Machine)
+    $cmd = "C:\k\kubelet.exe $global:KubeletArgs --cert-dir=$env:SYSTEMDRIVE\var\lib\kubelet\pki --config=/var/lib/kubelet/config.yaml --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --hostname-override=$(hostname) --pod-infra-container-image=`"mcr.microsoft.com/oss/kubernetes/pause:1.4.1`" --enable-debugging-handlers --cgroups-per-qos=false --enforce-node-allocatable=`"`" --network-plugin=cni --resolv-conf=`"`" --log-dir=/var/log/kubelet --logtostderr=false --image-pull-progress-deadline=20m"
 
-[Environment]::SetEnvironmentVariable("PATH", $newPath, [EnvironmentVariableTarget]::Machine)
+    Invoke-Expression $cmd'
+    $StartKubeletFileContent = $StartKubeletFileContent -replace "{{CONTAINER_RUNTIME}}", "`"$ContainerRuntime`""
+    Set-Content -Path $global:StartKubeletScript -Value $StartKubeletFileContent
 
-Write-Host "Registering kubelet service"
-nssm install kubelet $global:Powershell $global:PowershellArgs $global:StartKubeletScript
+    Write-Host "Registering kubelet service"
+    nssm install kubelet $global:Powershell $global:PowershellArgs $global:StartKubeletScript
 
-if ($ContainerRuntime -eq "Docker") {
-    nssm set kubelet DependOnService docker
-} elseif ($ContainerRuntime -eq "containerD") {
-    nssm set kubelet DependOnService containerd
+    if ($ContainerRuntime -eq "Docker") {
+        nssm set kubelet DependOnService docker
+    } elseif ($ContainerRuntime -eq "containerD") {
+        nssm set kubelet DependOnService containerd
+    }
+} else {
+    Write-Host "kubelet has been already installed"
 }
 
-New-NetFirewallRule -Name kubelet -DisplayName 'kubelet' -Enabled True -Direction Inbound -Protocol TCP -Action Allow -LocalPort 10250
+New-NetFirewallRule -Name kubelet -DisplayName 'kubelet' -Enabled True -Direction Inbound -Protocol TCP -Action Allow -LocalPort 10250 -ErrorAction Ignore
```

Comment:
How about writing it like this? This has the benefit of not catching unexpected errors.
```suggestion
if (-not [bool](Get-NetFirewallRule -Name "kubelet")) {
    New-NetFirewallRule -Name kubelet -DisplayName 'kubelet' -Enabled True -Direction Inbound -Protocol TCP -Action Allow -LocalPort 10250
}
```

Reply:
You are right. It is better.